cilium/CFP-27752-operator-manages-cilium-identities.md

```diff
@@ -0,0 +1,220 @@
+
+
+
+# Migration
+
+Any situation when neither Operator nor Agents are managing CIDs is effectively a downtime for assigning new Security Identities. Either of these components must be available and healthy in order to:
+- Deploy workloads with a new set of identity-relevant labels
+- Update existing workloads with different identity-relevant labels
+- Modify the identity-relevant labels filter
+- Modify namespace labels and enforce policies for the newly changed labels
+
+
+## Upgrade
+
+**From Agents managing CIDs to Operator managing CIDs”**
+
+Safe upgrade, without downtime, requires simultaneous CID management compatibility. The CID controller in Operator is designed to support simultaneous CID management and won’t interfere with Agents.
+
+The user needs to update Operator first to enable Operator managing CIDs, before Agents are updated to stop managing CIDs (the same flag `operator-manages-identities`)
+
+**From Operator managing CIDs to Operator managing CIDs”**
+
+No action is required from the user, because there won’t be a change in the CID management configuration.
+
+
+
+## Downgrade
+
+Downgrade without downtime is possible only if Agent daemonset is downgraded first -- restart after configuration change, or updated to an earlier version where Agent manages CIDs. Otherwise, if Operator is downgraded first, there will be a downtime lasting at least the amount of time needed for Agent daemonset to be downgraded after the Operator downgrade.
+
+# Security
+
+* Agents will no longer need to have the RBAC permission for modifying CIDs.
+
+
+# Future enhancements
+
+
+## CID lazy creation
+
+Scalability and performance optimization
+
+Create CIDs only for labels used in network policies to greatly reduce the number of CIDs. Only pod labels used in the peer pod label selector of network policies will be relevant for CID creation.
+
+There are multiple levels of optimization and complexity this solution can go to. The proposal is to start with a basic implementation that will have effects like CID relevant labels, based on label keys from network policy selectors. Discussion of more advanced solutions is outside the scope of this document.
+
+
+## Operator acts on Pod instead of CEP changes
+
+Status: Included in the original proposal
+
+Performance optimization
+
+Operator doesn't watch for CEPs but watches for Pods instead. As soon as a pod is created (not started) Operator allocates a CID immediately. This way the new CID is getting created in parallel as the pod is getting scheduled.
+
+Downside: This is a very minor downside and isn't expected to cause any problems. In case pods are unschedulable we will be allocating identities even though they might not be used at the end.
+
+## Deprecate CEP
+
+This topic entails the possibility of removing CEPs altogether, because we can just have CIDs and CESs. Agents don’t create CEPs anymore, just watch for CESs and CIDs. Operator creates both CIDs and CESs with all info gathered from pods.
```

Comment:
is this planned for initial release or for future?

Reply:
This milestone is under a section called "Future enhancements". I consider this to be subject to future discussion. None of the items under this section are considered "Implementable" per the intended status of the CFP, these are ideas that can help guide this CFP and may be designed/implemented in future.